src/ray/raylet/node_manager.cc

```diff
@@ -2793,7 +2776,11 @@ void NodeManager::TriggerGlobalGC() {
   should_local_gc_ = true;
 }
 
-void NodeManager::Stop() { object_manager_.Stop(); }
+void NodeManager::Stop() {
+  object_manager_.Stop();
+  dashboard_agent_manager_.reset();
```

Comment:
Is .reset necessary? Why don't we just make them a none-pointer and not care about reset instead? (I've seen several cases early .reset + destructor logic segfault upon shutdown path)

Reply:
It did not work. AgentManager needs to be constructed before the node manager ctor function body, but the agent needs to know the node manager's port which is assigned in node_manager_server_.Run() in the ctor body. We can init an empty AgentManager and assign later but that adds a lot of unnecessary complexity so I reverted.

To reduce cognitive cost, I changed shared_ptr to unique_ptr.

Reply:
If we did that (construct the agent manager before node_manager_server starts) we will get a 0 port and it causes dashboard test to fail (which listens on this port) https://buildkite.com/ray-project/oss-ci-build-pr/builds/29201#01897aca-7ffe-44df-ac09-496ee5d4af9d

Reply:
unique_ptr sgtm